File: solutions/java/problems/[0316_Medium] Remove Duplicate Letters.java
class Solution {
    public String removeDuplicateLetters(String s) {
        int n = s.length();
        int[] lastPos = new int[26];
        for (int i = 0; i < n; i++) {
            char c = s.charAt(i);
            lastPos[c - 'a'] = i;
        }
        Stack<Character> stack = new Stack<>();
        Set<Character> charSet = new HashSet<>();
        for (int i = 0; i < n; i++) {
            char c = s.charAt(i);
            if (!charSet.contains(c)) {
                while (!stack.empty() && stack.peek() > c && lastPos[stack.peek() - 'a'] > i) {
                    charSet.remove(stack.peek());
                    stack.pop();
                }
                charSet.add(c);
                stack.push(c);
            }
        }
        StringBuilder ans = new StringBuilder();
        for (Character c : stack) ans.append(c.charValue());
        return ans.toString();
    }
}

File: solutions/c++/problems/[0316_Medium] Remove Duplicate Letters.cpp
class Solution {
public:
    string removeDuplicateLetters(string s) {
        int n = s.size();
        vector <vector<int>> q(26);
        for (int i = 0; i < n; i++) q[s[i] - 'a'].push_back(i);
        string ans = "";
        unordered_set<char> charSet;
        for (int i = 0; i < n; i++) {
            char c = s[i];
            if (charSet.count(c) == 0) {
                while (!ans.empty() && ans.back() > c && q[ans.back() - 'a'].back() > i) {
                    charSet.erase(ans.back());
                    ans.pop_back();
                }
                charSet.insert(c);
                ans += c;
            }
        }
        return ans;
    }
};